### PR TITLE
Removed Hard-Coded Default Log's Pagination Configuration, instead va…

### PR DESCRIPTION
**Pull request by N3XT14**
## Overview

The proposed change in this pull request involves removing the hard-coded default log pagination configuration and instead fetching values from settings.py. This change will make the code more flexible and easier to maintain, as users can easily modify the log pagination configuration without needing to modify any code.

### Example
```python
from prefect.settings import (
    PREFECT_API_DEFAULT_LOGS_PAGE_SIZE,
    PREFECT_API_DEFAULT_NUM_LOGS,
)

LOGS_DEFAULT_PAGE_SIZE = PREFECT_API_DEFAULT_LOGS_PAGE_SIZE.value()
LOGS_WITH_LIMIT_FLAG_DEFAULT_NUM_LOGS = PREFECT_API_DEFAULT_NUM_LOGS.value()
```
```sh
$ prefect flow-run logs c1b58559-cae1-400e-bb87-6b3827934e95
2023-04-06 16:27:02.656 | INFO    | Flow run 'mellow-antelope' - Created task run 'call_api-0' for task 'call_api'
2023-04-06 16:27:02.657 | INFO    | Flow run 'mellow-antelope' - Executing 'call_api-0' immediately...
2023-04-06 16:27:03.158 | INFO    | Flow run 'mellow-antelope' - Finished in state Completed()
2023-04-06 16:27:03.209 | INFO    | Flow run 'mellow-antelope' - Created task run 'parse_fact-0' for task 'parse_fact'
2023-04-06 16:27:03.210 | INFO    | Flow run 'mellow-antelope' - Executing 'parse_fact-0' immediately...
2023-04-06 16:27:03.349 | INFO    | Flow run 'mellow-antelope' - Finished in state Completed()
2023-04-06 16:27:03.410 | INFO    | Flow run 'mellow-antelope' - Finished in state Completed()

$ prefect flow-run logs c1b58559-cae1-400e-bb87-6b3827934e95 --head
2023-04-06 16:27:02.656 | INFO    | Flow run 'mellow-antelope' - Created task run 'call_api-0' for task 'call_api'
2023-04-06 16:27:02.657 | INFO    | Flow run 'mellow-antelope' - Executing 'call_api-0' immediately...
2023-04-06 16:27:03.158 | INFO    | Flow run 'mellow-antelope' - Finished in state Completed()
2023-04-06 16:27:03.209 | INFO    | Flow run 'mellow-antelope' - Created task run 'parse_fact-0' for task 'parse_fact'
2023-04-06 16:27:03.210 | INFO    | Flow run 'mellow-antelope' - Executing 'parse_fact-0' immediately...
2023-04-06 16:27:03.349 | INFO    | Flow run 'mellow-antelope' - Finished in state Completed()
2023-04-06 16:27:03.410 | INFO    | Flow run 'mellow-antelope' - Finished in state Completed()
```

### Checklist
- [x] This pull request references any related issue by including "#8002`"
- [ ] This pull request includes tests or only affects documentation.
- [x] This pull request includes the enhancement label.

### CODE REVIEW

**Comment by madkinsz** [comment]
Hey @N3XT14 — sorry for the lack of clarity in the issue but this isn't quite what I meant. What we want to do is:

1. Ask for the first page of logs without specifying a limit (unless, one has been specified as a flag)
2. Look at the number of logs returned and treat _that_ as the page size for the next request
3. If we get less than the number of logs in a page, we know we've reached the end and can stop

Relevant section is at https://github.com/PrefectHQ/prefect/blob/main/src/prefect/cli/flow_run.py#L199-L215

We can continue to use a hard-coded value of 20 for the default short logs view. 

Does this make sense?

**Comment by N3XT14** [comment]
@madkinsz No Issues, Thanks for the clarification and before I start I want to see if the understanding was clear to me.
Let me know if I understood it correctly and the changes that I intend to make

Q1: Ask for the first page of logs without specifying a limit (unless one has been specified as a flag)
Meaning, making a client request to fetch the logs without the limit option or set to None if not specified explicitly.

What do you mean by the first page of logs?
Is it making a client request and if yes so we need to specify any option in the API read_logs such that it returns only the first page and what would that option be because in the API I can't see any option as page no.

Q2: Look at the number of logs returned and treat that as the page size for the next request
Here targeted variable is "num_logs_to_return_from_page". 

Q3: If we get less than the number of logs on a page.
In this, the comparison is made between the count of the num of logs returned by the client request of the first page with?
Asking this because if the count if less then I can make more_logs false so it won't enter while and will just print the logs for one time.

Any feedback is appreciated. Do let me know if I am missing something maybe due to a beginner contributor or lack of understanding.

**Comment by madkinsz** [comment]
> Q1: Ask for the first page of logs without specifying a limit (unless one has been specified as a flag) Meaning, making a client request to fetch the logs without the limit option or set to None if not specified explicitly.
> 
> What do you mean by the first page of logs? Is it making a client request and if yes so we need to specify any option in the API read_logs such that it returns only the first page and what would that option be because in the API I can't see any option as page no.

The API has a default limit that will be applied if one is not given. Practically, this means that the API will return 200 logs on the first request if a limit is not given and we can assume that is the size of a "page". The problem is that the API can be configured to have a different limit so we can't just always assume it's 200.

> Q2: Look at the number of logs returned and treat that as the page size for the next request Here targeted variable is "num_logs_to_return_from_page".

Yes. That variable name isn't super clear to be honest. It could just be "page_size".

> Q3: If we get less than the number of logs on a page. In this, the comparison is made between the count of the num of logs returned by the client request of the first page with? Asking this because if the count if less then I can make more_logs false so it won't enter while and will just print the logs for one time.

I'm a bit confused by this one. If we get less logs back than the requested limit then we have reached the end of the logs and should stop.

**Comment by N3XT14** [comment]
@madkinsz . Thanks for the clarification once again. I have done the required changes and would love to hear your feedback on it.

To ease the changes, I am providing a simulation of the approach which reflects in the code.
Firstly, I am making a client request which provides the page_size limit set by the API i.e when the limit is not set by any flags.

That limit serves as a page_size for the rest of the request and accordingly, all the changes are reflected in the following code.
Let me know whether the proposed solution is what you meant.

I have tested it so that it doesn't break the existing functionality and below is a small snippet of the same.
For the purpose to test, I have printed length instead of actual logs. So accordingly, if the API limit is increased the code will pick that limit to serve the request.

```.sh
$ prefect flow-run logs 543c6411-9f2d-4c12-83e5-b4a5a2f3fdc5 --head --num-logs 100
Num of logs returned =>  100
Page Size 200
(.venv)
Yash Oswal@N3XT MINGW64 /d/ME/Projects/Open Source/prefect (fix-logging)
$ prefect flow-run logs 543c6411-9f2d-4c12-83e5-b4a5a2f3fdc5 --head
Num of logs returned =>  20
Page Size 200
(.venv)
Yash Oswal@N3XT MINGW64 /d/ME/Projects/Open Source/prefect (fix-logging)
$ prefect flow-run logs 543c6411-9f2d-4c12-83e5-b4a5a2f3fdc5
Num of logs returned =>  200
Page Size 200
Num of logs returned =>  211
Page Size 200
```

I have changed some of the variables so that it reduced a bit of confusion in understanding. 
Also, on the first request, I am not changing the offset meaning the request is made again which could be avoided but would require more changes. So, let me know whether that should be incorporated or not.

**Comment by madkinsz** [comment]
@N3XT14 this looks good!

> Also, on the first request, I am not changing the offset meaning the request is made again which could be avoided but would require more changes. So, let me know whether that should be incorporated or not.

It seems worth avoiding a duplicate API request here. I think you could do this easily by moving subsequent log retrieval to the end of the loop instead of the start?

File: src/prefect/settings.py [hunk 1]
```diff
@@ -1015,6 +1015,20 @@ def default_cloud_ui_url(settings, value):
...
+)
+"""The default page size applied to the logs`.
+"""
+
+PREFECT_API_DEFAULT_NUM_LOGS = Setting(
+    int,
+    default=20,
+)
+"""he default num of logs applied to arguments head or num_logs`.
+"""
```
Comment (madkinsz) [inline src/prefect/settings.py:1030]: Let's remove these since we don't need them

File: src/prefect/cli/flow_run.py [hunk 2]
```diff
@@ -22,15 +22,19 @@
...
 
 flow_run_app = PrefectTyper(
     name="flow-run", help="Commands for interacting with flow runs."
 )
 app.add_typer(flow_run_app, aliases=["flow-runs"])
 
-LOGS_DEFAULT_PAGE_SIZE = 200
-LOGS_WITH_LIMIT_FLAG_DEFAULT_NUM_LOGS = 20
+LOGS_DEFAULT_PAGE_SIZE = PREFECT_API_DEFAULT_LOGS_PAGE_SIZE.value()
+LOGS_WITH_LIMIT_FLAG_DEFAULT_NUM_LOGS = PREFECT_API_DEFAULT_NUM_LOGS.value()
```
Comment (madkinsz) [inline src/prefect/cli/flow_run.py:37]: Let's leave this hard-coded to 20

**Comment by N3XT14** [comment]
@madkinsz Changes are done as needed. I cleaned up the code, and moving the log retrieval at the end is what needs to be done.

For the purpose to test, I have printed length instead of actual logs.
```sh
Yash Oswal@N3XT MINGW64 /d/ME/Projects/Open Source/prefect (fix-logging)
$ prefect flow-run logs 543c6411-9f2d-4c12-83e5-b4a5a2f3fdc5 --head --num-logs 250
Num of logs returned =>  200
Page Size =>  200
Num of logs returned =>  211
Page Size =>  200
(.venv) 
Yash Oswal@N3XT MINGW64 /d/ME/Projects/Open Source/prefect (fix-logging)
$ prefect flow-run logs 543c6411-9f2d-4c12-83e5-b4a5a2f3fdc5 --head --num-logs 100
Num of logs returned =>  100
Page Size =>  200
(.venv) 
Yash Oswal@N3XT MINGW64 /d/ME/Projects/Open Source/prefect (fix-logging)
$ prefect flow-run logs 543c6411-9f2d-4c12-83e5-b4a5a2f3fdc5 --head
Num of logs returned =>  20
Page Size =>  200
(.venv) 
Yash Oswal@N3XT MINGW64 /d/ME/Projects/Open Source/prefect (fix-logging)
$ prefect flow-run logs 543c6411-9f2d-4c12-83e5-b4a5a2f3fdc5
Num of logs returned =>  200
Page Size =>  200
Num of logs returned =>  211
Page Size =>  200
(.venv) 
```

Thanks

**Comment by serinamarie** [comment]
Hi @N3XT14, we just merged some changes into main that affect one of your changed files. It would be best for you to pull them to your branch and resolve any conflicts! Also, it would be great to see some tests to ensure we are inferring from settings instead of hard-coding. Let me know if you have any questions :)

**Comment by madkinsz** [comment]
@N3XT14 Sometimes if the history diverges it'll break the PR — let me know if you need help rebasing!

**Comment by N3XT14** [comment]
@madkinsz Yup. any help is appreciated. Let me know what steps should I follow.

FYI, my fork/branch is updated/sync with the main branch and almost resolved all the conflicts but PR has been closed

**Comment by N3XT14** [comment]
@serinamarie 

I am not actually referring to the settings.py file but making an initial API call to get the page size before entering the loop as discussed with @madkinsz above.  This acts as an upper bound of num of logs to be returned. So if API has increased the limit to num of logs to be sent on each request when no limits are specified then accordingly the default page size will be set. This will ensure that the page size/default size is dynamic.

As for tests, I created a sample test file where I created a task and made calls that generate more than 200 logs. With this, I am testing all the scenarios with all the command line options.

Let me know if I am on the right track and apologies as I am taking quite a long time for a trivial task like this.

**Comment by serinamarie** [comment]
@N3XT14 I'm unable to view any file changes unfortunately.

> FYI, my fork/branch is updated/synced with the main branch and almost resolved all the conflicts but PR has been closed which I cannot reopen due to permission.

I'm also unable to reopen the PR 🤔 Maybe @madkinsz can?

**Comment by N3XT14** [comment]
@madkinsz Are you able to reopen this PR? If not, @serinamarie Should I open a new PR?

**Pull request closed without merging**